src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,929 @@
+    - Since r is larger than lo the bit must be 0 in lo and 1 in r
+    - Since r must satisify bits the bit must be 0 in zeros
+    - Since r should be the smallest value, this bit should be the lowest one
+      possible
+
+  E.g:      1 2 3 4 5 6
+       lo = 1 0 0 1 1 0
+        x = 1 0 1 0 1 0
+        y = 0 1 1 1 1 1
+  x would be larger than lo since the first different bit is the 3rd one,
+  while y is smaller than lo because the first different bit is the 1st bit.
+  Next, consider:
+       x1 = 1 0 1 0 1 0
+       x2 = 1 0 0 1 1 1
+  Both x1 and x2 are larger than lo, but x1 > x2 since its first different
+  bit from lo is the 3rd one, while with x2 it is the 7th one. As a result,
+  if both x1 and x2 satisfy bits, x2 would be closer to our true result.
+
+  2. Formality:
+
+  Call r the smallest value not smaller than lo that satisfies bits. Since lo
+  does not satisfy bits, lo < r (2.1)
+
+  Call i the largest bit index such that:
+
+  - lo[x] satisfies bits for 0 <= x < i (2.2)
+  - zeros[i] = 0                        (2.3)
+  - lo[i]    = 0                        (2.4)
+
+  Consider v:
+
+  - v[x] = lo[x], for 0 <= x < i        (2.5)
+  - v[i] = 1                            (2.6)
+  - v[x] = ones[x], for x > i           (2.7)
+
+  We will prove that v == r.
+
+  a. Firstly, we prove that r <= v:
+
+    a.1. We have lo < v since:
+      lo[x] == v[x], for 0 <= x < i (according to 2.5)
+      lo[i] <  v[i] (according to 2.4 and 2.6, lo[i] == 0 < v[i] == 1)
+      bits at x > i have lower significance, and are thus irrelevant
+
+    a.2. We have v satisfies bits, this is because:
+      v[x] satisfies bits for 0 <= x < i (according to 2.2 and 2.5)
+      v[i] satisfies bits:
+        According to 2.3 and 2.6, v[i] == 1 and zeros[i] == 0, v[i] does not violate
+        bits, which means v[i] satisfies bits
+      v[x] satisfies bits for x > i:
+        Assume bits is not contradictory, we cannot have:
+          ones[x]  == 1, v[x] == 0 (according to 2.7, v[x] == ones[x])
+          zeros[x] == 1, v[x] == 1 (according to 2.7, ones[x] == v[x] == 1, which means
+                                    bits is contradictory)
+
+    From a.1 and a.2, v > lo and v satisfies bits. Which means r <= v since r is the
+    smallest such value.
+
+  b. Secondly, from r <= v, we prove that r == v. Suppose the contradiction r < v:
+
+    Since r < v, there must be a bit position j that:
+
+    r[j] == 0               (2.b.1)
+    v[j] == 1               (2.b.2)
+    r[x] == v[x], for x < j (2.b.3)
+
+    b.1. If j < i
+      This means that:
+      r[j]  == 0                (according to 2.b.1)
+      lo[j] == 1                (according to 2.b.2 and 2.5, lo[j] == v[j] == 1 because j < i)
+      r[x]  == lo[x], for x < j (according to 2.b.3 and 2.5, lo[x] == v[x] == r[x] with x < j < i)
+      bits at x > j have lower significance, and are thus irrelevant
+
+      Which leads to r < lo, which contradicts that r >= lo
```

Comment:
```suggestion
      Which leads to r < lo, which contradicts that r >= lo (according to definition of r)
```
It could be nice to define it more explicitly, and give the definition of `r` a title / name.

Reply:
We can say according to 2.1 here, done that!